Add content for local tools (#17065)
dotnet/docs: docs/core/tools/global-tools.md

---
# How to manage .NET Core tools

**This article applies to:** ✔️ .NET Core 2.1 SDK and later versions

A .NET Core tool is a special NuGet package that contains a console application. A tool can be installed on your machine in the following ways:

* As a global tool.

  The tool binaries are installed in a default directory that is added to the PATH environment variable. You can invoke the tool from any directory on the machine without specifying its location. One version of a tool is used for all directories on the machine.

* As a global tool in a custom location (also known as a tool-path tool).

  The tool binaries are installed in a location that you specify. You can invoke the tool from the installation directory or by providing the directory with the command name or by adding the directory to the PATH environment variable. One version of a tool is used for all directories on the machine.

* As a local tool (applies to .NET Core SDK 3.0 and later).

  The tool binaries are installed in a default directory. You invoke the tool from the installation directory or any of its subdirectories. Different directories can use different versions of the same tool.
  

>   The .NET CLI uses manifest files to keep track of which tools are installed as local to a directory. These manifest files are saved in the source code repository, so they're available when a contributor checks out the code from the repository. The contributor can then invoke a single .NET Core CLI command that installs all of the tools that are listed in the manifest files.

Reviewer: Changing this. It made it sound like this was the only scenario you would use this, but it may not be.

```suggestion
  The .NET CLI uses manifest files to keep track of which tools are installed as local to a directory. When the manifest file is saved in the root directory of a source code repository, a contributor can clone the repository and invoke a single .NET Core CLI command that installs all of the tools listed in the manifest files.
```

```

The .NET Core SDK doesn't add this location automatically to the PATH environment variable. To invoke a tool-path tool, you have to make sure the command is available by using one of the following methods:

* Add the installation directory to the PATH environment variable.
* Specify the full path to the tool when you invoke it.
* Invoke the tool from within the installation directory.

## Install a local tool

**Applies to .NET Core 3.0 SDK and later.**

To install a tool for local access only (for the current directory and subdirectories), it has to be added to a tool manifest file. To create a tool manifest file, run the `dotnet new tool-manifest` command:

```dotnetcli
dotnet new tool-manifest
```

This command creates a manifest file named *dotnet-tools.json* under the *.config* directory. To add a local tool to the manifest file, use the [dotnet tool install](dotnet-tool-install.md) command and **omit** the `--global` and `--tool-path` options, as shown in the following example:

```dotnetcli
dotnet tool install dotnetsay
```

The command output shows which manifest file the newly installed tool is in, as shown in the following example:

> ```

Reviewer: Example after this has a `console` code language, but this one doesn't.

You can invoke the tool from this directory using the following command:
dotnet tool run dotnetsay
Tool 'dotnetsay' (version '2.1.4') was successfully installed.

> Entry is added to the manifest file /Users/name/tools/sub/dotnet-tools.json

Reviewer: The output here demonstrates linux/mac pathing, but the example below for the `dotnet tool list` command uses windows pathing.

```console
Tool 'botsay' (version '1.0.0') was restored. Available commands: botsay
Tool 'dotnetsay' (version '2.1.3') was restored. Available commands: dotnetsay
Restore was successful.
```

## Install a specific tool version

To install a pre-release version or a specific version of a tool, specify the version number by using the `--version` option, as shown in the following example:

```dotnetcli
dotnet tool install dotnetsay --version 2.1.3
```

## Use a tool

The command that you use to invoke a tool may be different from the name of the package that you install. To display all of the tools currently installed on the machine for the current user, use the [dotnet tool list](dotnet-tool-list.md) command:

```dotnetcli
dotnet tool list
```

The output shows each tool's version and command, similar to the following example:

```console
Package Id      Version      Commands       Manifest
-------------------------------------------------------------------------------------------
botsay          1.0.0        botsay         /home/name/repository/.config/dotnet-tools.json
dotnetsay       2.1.3        dotnetsay      /home/name/repository/.config/dotnet-tools.json
```

> As shown in this example, the list shows local tools. To see global tools, use the `--global` option, and to see tool-path global tools, use the `--tool-path` option.

Author: Standardized as "tool-path tool"

* A global tool and a local tool have the same command prefixed by `dotnet-`.
* You want to invoke the global tool from a directory that is in scope for the local tool.

In this scenario, `dotnet doc` and `dotnet dotnet-doc` invoke the local tool. To invoke the global tool, use the command by itself:

```dotnetcli
dotnet-doc
```

### Invoke a tool-path global tool

To invoke a global tool that is installed by using the `tool-path` option, make sure the command is available, as explained [earlier in this article](#install-a-global-tool-in-a-custom-location).

### Invoke a local tool

To invoke a local tool, you have to use the `dotnet` command from within the installation directory. You can use the long form (`dotnet tool run <tool>`) or the short form (`dotnet <tool>`), as shown in the following examples:

```dotnetcli
dotnet tool run dotnetsay
dotnet dotnetsay
```

If the command is prefixed by `dotnet-`, include that prefix when you invoke the tool. For example, if the command is `dotnet-doc`, either of the following examples invokes the local tool:

```dotnetcli
dotnet tool run dotnet-doc

> dotnet dotnet-doc

Reviewer 2: and `dotnet doc`?

### Install a global tool in a custom location

To install a tool as a global tool in a custom location, use the `--tool-path` option of [dotnet tool install](dotnet-tool-install.md), as shown in the following examples.

On Windows:

```dotnetcli
dotnet tool install dotnetsay --tool-path c:\dotnet-tools
```

On Linux or macOS:

```dotnetcli
dotnet tool install dotnetsay --tool-path ~/bin
```

If installation is successful, a message is displayed showing the command used to invoke the tool and the version installed, similar to the following example:

```output
You can invoke the tool using the following command: dotnetsay
Tool 'dotnetsay' (version '3.0.0') was successfully installed.
```

The .NET Core SDK doesn't add this location automatically to the PATH environment variable. To invoke a tool-path tool, you have to make sure the command is available by using one of the following methods:

* Add the installation directory to the PATH environment variable.
* Specify the installation directory when you invoke the tool.
* invoke the tool from within the installation directory.

### Install a local tool

> **Applies to .NET Core 3.0 and later.**

Reviewer 3: ```suggestion
**Applies to .NET Core 3.0 SDK and later.**
```

```dotnetcli
dotnet tool restore
```

The output indicates which tools were restored:

```console
Tool 'botsay' (version '1.0.0') was restored. Available commands: botsay
Tool 'dotnetsay' (version '2.1.3') was restored. Available commands: dotnetsay
Restore was successful.
```

## Install a specific tool version

To install a pre-release version or a specific version of a tool, specify the version number by using the `--version` option, as shown in the following example:

```dotnetcli
dotnet tool install dotnetsay --version 2.1.3
```

## Use a tool

The command that you use to invoke a tool may be different from the name of the package that you install. To display all of the tools currently installed on the machine for the current user, use the [dotnet tool list](dotnet-tool-list.md) command:

```dotnetcli
dotnet tool list
```

The output shows each tool's version and command, similar to the following example:

```console
Package Id      Version      Commands       Manifest
------------------------------------------------------------------------------------------------------
botsay          1.0.0        botsay         C:\Users\username\Documents\botsay\.config\dotnet-tools.json

> dotnetsay       2.1.3        dotnetsay      C:\Users\username\Documents\botsay\.config\dotnet-tools.json

Reviewer 4: It seems odd that the manifest files is buried in the botsay directory because it's probably not a parent directory of where dotnetsay would be run from?

Author: Where you run a command from is determined by where you install it.  Since in this tutorial you install dotnetsay in botsay, that's where you run it from.  I could expand the scenario by installing dotnetsay in a different directory.  Do you think that is required here to make clear how it works?

Reviewer 4: I think it's a little weird that the dotnetsay tool lives in the botsay folder. Is there a way to install them in a "Tools" folder? That would seem to make more sense to me.

Author: The tools folder installation sounds more appropriate to a global tool in custom location, aka tool-path tool. Local tools are tools you want to use with one project so you can do things like use a different version of the same tool with different projects.

Reviewer: I agree. It's confusing to have the folder named botsay (the name of an existing tool). I would use something like `project` or `code` or `repository`

Author: How about this:  start the first tutorial by creating a "repository" folder, and create botsay under that, then in the local tools tutorial install the tools in the repository folder.

```

As shown in this example, the list shows local tools. To see global tools, use the `--global` option, and to see tool-path global tools, use the `--tool-path` option.

### Invoke a global tool

For global tools, use the tool command by itself. For example:

If the command is `dotnetsay`, that's what you use to invoke the command:

```console
dotnetsay
```

If the tool author specified a command with the prefix `dotnet-`, you can invoke the tool with or without the `dotnet` command. For example, if the command is `dotnet-doc`, you can invoke the tool in either of the following ways:

```dotnetcli
dotnet-doc

> dotnet doc

Author: You can invoke it with dotnet-doc without using dotnet in front of it.

> dotnet doc

Reviewer 4: Perhaps we should add a note here and/or in the "invoke a local tool" section that the dotnet- prefix is mandatory for local tools but not global tools, just to be perfectly clear. It seems like an arbitrary rule, so might be worth clarifying.

> dotnet doc

Reviewer 4: Also, the example keeps confusing me because it seems to contradict with the text above it. Is it that the dotnet- prefix isn't required, or that the dotnet command isn't required?

```

However, if a global tool and a local tool have the same name prefixed by `dotnet-` and you want to invoke the global tool, you have to include the prefix. If the command is `dotnet-doc` for both a global and a local tool,`dotnet doc` invokes the local tool.

### Invoke a tool-path global tool

To invoke a global tool that is installed by using the `tool-path` option, make sure the command is available by using one of the following methods:

* Add the installation directory to the PATH environment variable.
* Specify the installation directory when you invoke the tool.
* invoke the tool from within the installation directory.

### Invoke a local tool

To invoke a local tool, use the `dotnet` command or the `dotnet tool run` command from within the installation directory, as shown in the following example:

```dotnetcli
dotnet tool run dotnetsay
dotnet dotnetsay
```

If the command is prefixed by `dotnet-`, include that prefix when you invoke the tool. For example, if the command is `dotnet-doc`, either of the following examples invokes the local tool:

```dotnetcli
dotnet tool run dotnet-doc
dotnet dotnet-doc
```

> ## Update or uninstall a tool

Author: I'll make get help an h2 so this h2 only has update and uninstall
